src/metrics/src/lib.rs

```diff
@@ -17,13 +17,166 @@
 
 #![warn(missing_docs, missing_debug_implementations)]
 
+use std::time::Duration;
+
+use mz_dyncfg::{ConfigSet, ConfigUpdates};
 use mz_ore::metrics::MetricsRegistry;
+use tokio::time::Interval;
+use tokio_util::sync::CancellationToken;
+
+pub use dyncfgs::all_dyncfgs;
 
+mod dyncfgs;
 pub mod lgalloc;
 pub mod rusage;
 
+/// Handle to metrics defined in this crate.
+#[derive(Debug)]
+pub struct Metrics {
+    config_set: ConfigSet,
+    lgalloc: MetricsTask,
+    rusage: MetricsTask,
+}
+
+static METRICS: std::sync::Mutex<Option<Metrics>> = std::sync::Mutex::new(None);
+
 /// Register all metrics into the provided registry.
-pub async fn register_metrics_into(metrics_registry: &MetricsRegistry) {
-    lgalloc::register_metrics_into(metrics_registry).await;
-    rusage::register_metrics_into(metrics_registry).await;
+///
+/// This function is async, because it needs to be called from a tokio runtime context.
+#[allow(clippy::unused_async)]
+pub async fn register_metrics_into(metrics_registry: &MetricsRegistry, config_set: ConfigSet) {
+    let update_duration_metric = metrics_registry.register(mz_ore::metric!(
+        name: "mz_metrics_update_duration",
+        help: "The time it took to update lgalloc stats",
+        var_labels: ["name"],
+        buckets: mz_ore::stats::histogram_seconds_buckets(0.000_500, 32.),
+    ));
+
+    let lgalloc = Metrics::new_metrics_task(
+        metrics_registry,
+        lgalloc::register_metrics_into,
+        dyncfgs::MZ_METRICS_LGALLOC_REFRESH_INTERVAL,
+        &update_duration_metric,
+    );
+    let rusage = Metrics::new_metrics_task(
+        metrics_registry,
+        rusage::register_metrics_into,
+        dyncfgs::MZ_METRICS_RUSAGE_REFRESH_INTERVAL,
+        &update_duration_metric,
+    );
+
+    *METRICS.lock().unwrap() = Some(Metrics {
+        lgalloc,
+        rusage,
+        config_set,
+    })
+}
+
+/// Update the configuration of the metrics.
+pub fn update_dyncfg(config_updates: &ConfigUpdates) {
+    if let Some(metrics) = METRICS.lock().expect("lock poisoned").as_mut() {
+        config_updates.apply(&metrics.config_set);
+        metrics.apply_dyncfg_updates();
```

Comment:
IIUC the ordering of these function calls is important? It would be nice to call this out because at first glance it's not immediately obvious that the second function call is dependent on the first

Reply:
@antiguru I'm going to merge this, but mind following up here if my understanding is correct?